src/algebra/free_monoid.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2019 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Simon Hudon, Yury Kudryashov
+-/
+import algebra.group.hom data.equiv.algebra data.list.basic
+
+/-!
+# Free monoid over a given alphabet
+
+## Main definitions
+
+* `free_monoid α`: free monoid over alphabet `α`; defined as a synonym for `list α`
+  with multiplication given by `(++)`.
+* `free_monoid.of`: embedding `α → free_monoid α` sending each element `x` to `[x]`;
+* `free_monoid.lift α M`: natural equivalence between `α → M` and `free_monoid α →* M`;
+  for technical reasons `α` and `M` are explicit arguments;
+* `free_monoid.map`: embedding of `α → β` into `free_monoid α →* free_monoid β` given by `list.map`.
+-/
+
+variables {α : Type*} {β : Type*} {γ : Type*} {M : Type*} [monoid M]
+
+/-- Free monoid over a given alphabet. -/
+@[to_additive free_add_monoid "Free additive monoid over a given alphabet"]
```

Comment:
```suggestion
@[to_additive free_add_monoid "Free nonabelian additive monoid over a given alphabet"]
```
In general people think that additive stuff is abelian, so adding a precision can not hurt.